### DISCUSSION TITLE
Normal cost function docs question

### DISCUSSION THREAD

**Discussion by Rotario** [deepcharles/ruptures#127] [Q&A] [answered]
Copied this from Issues as requested

Hi guys,
Love your work!
I'd just like to clarify something with the Normal cost function as mentioned in the User guide here
It says
"This cost function detects changes in the mean and covariance matrix of a sequence of multivariate Gaussian random variables"

But I can't see how the function detects changes in the mean?
This

cov = np.cov(sub.T)
_, val = slogdet(cov)
return val * (end - start)
Calculates the covariance and then its log.
I'm just wondering, where is the change in mean taken into account?

EDIT:
I can see Lavielle 2006 uses the same formula and mentions it detects changes in mean:
"For the detection of changes in the mean vector and/or the covariance matrix of a multivariate
sequence of random variables, this contrast also reduces to"
image

Thanks,
Rowan

**Comment by Rotario** [comment]
This is the answer I got from Olivier (@oboulant if you want to re-answer this here I'll delete this)

Hi Rowan,

Thanks a lot for your interest in ruptures !

Since this is a discussion rather than an issue, could you please move this to the discussion section ? I will answer here and close this issue in few days. I will also post the same answer in the discussion section if you create a new one.

In order to illustrate/understand/showcase how this cost function can capture as well changes in the mean, let's take an example where only the mean signal changes, for instance :

a 1D signal with 500 points
a single change point in the mean signal (around position 250)
the same variance on both segments of the signal (the noise standard deviation added to the constant signal has a value of 1
We explore three scenarios :

No change point
Now, if you consider no change point, then the cost function is computed on the whole signal and since you have a change in mean at around position 250, the computed variance on the complete signal is quite high.
![no_change_point](https://user-images.githubusercontent.com/29485059/109277793-af8c9280-780f-11eb-88cc-e3f01595721d.png)


One change point, at an non-optimal position
Let say you choose that there is a change point at position 100. Then, the cost function on the first segment will be relatively low because the variance will be low on this segment. The cost function on the second segment will be on the other hand relatively high since the real change point in the mean at around position 250 introduces some additional variance.
![non_optimal_change_point](https://user-images.githubusercontent.com/29485059/109277756-a4d1fd80-780f-11eb-94f7-c980b14ad182.png)

One change point, at optimal position
Let say you choose that the change point is at its real location. Then you are fine. Indeed, on both the first and second segments, the computed variances on each segment do not include the real change point at around position 250. Therefore, on both segments the computed variance (hence the cost) will be relatively low. The sum of costs on your segments is then at the minimum value possible.
![image](https://user-images.githubusercontent.com/29485059/109277746-a1d70d00-780f-11eb-8704-801655cf5233.png)

Please, find below a code sample you can run to illustrate this 🔽
```
import numpy as np
import matplotlib.pylab as plt
import ruptures as rpt

n, dim = 500, 1  # number of samples, dimension
n_bkps, sigma = 1, 1  # number of change points, noise standart deviation
signal, bkps = rpt.pw_constant(n, dim, n_bkps, noise_std=sigma, delta=(4, 5), seed=1234)

# Create cost object
c = rpt.costs.CostNormal().fit(signal)

# No change point
print("No change points:")
print(f"\tTotal cost:\t{c.error(0, 500):.2f}")
figure, axs = rpt.display(signal, bkps)
figure.show()

# 1 change point, at an non-optimal position
print("One non-optimal change points:")
print(f"\tCost on segment 1:\t{c.error(0, 100):.2f}")
print(f"\tCost on segment 2:\t{c.error(100, 500):.2f}")
print(f"\tTotal cost:\t{c.sum_of_costs([100, 500]):.2f}")
figure, axs = rpt.display(signal, bkps, [100, 500])
figure.show()

# 1 change point, at the optimal position
print("One optimal change points:")
print(f"\tCost on segment 1:\t{c.error(0, bkps[0]):.2f}")
print(f"\tCost on segment 2:\t{c.error(bkps[0], 500):.2f}")
print(f"\tTotal cost:\t{c.sum_of_costs(bkps):.2f}")
figure, axs = rpt.display(signal, bkps, bkps)
figure.show()
```
Hope I am clear enough and that you get the intuition behind it ? The logic is exactly the same for multidimensional signals and for several change points.

Best,

Olivier

**Comment by oboulant** [comment, marked as answer]
> Copied this from Issues as requested
> 
> Hi guys,
> Love your work!
> I'd just like to clarify something with the Normal cost function as mentioned in the User guide here
> It says
> "This cost function detects changes in the mean and covariance matrix of a sequence of multivariate Gaussian random variables"
> 
> But I can't see how the function detects changes in the mean?
> This
> 
> cov = np.cov(sub.T)
> _, val = slogdet(cov)
> return val * (end - start)
> Calculates the covariance and then its log.
> I'm just wondering, where is the change in mean taken into account?
> 
> EDIT:
> I can see Lavielle 2006 uses the same formula and mentions it detects changes in mean:
> "For the detection of changes in the mean vector and/or the covariance matrix of a multivariate
> sequence of random variables, this contrast also reduces to"
> image
> 
> Thanks,
> Rowan

[posting this in the discussion section, coming initially from an [issue](https://github.com/deepcharles/ruptures/issues/125#issuecomment-784961440) ]

Hi Rowan,

Thanks a lot for your interest in `ruptures` !

In order to illustrate/understand/showcase how this cost function can capture as well changes in the mean, let's take an example where only the mean signal changes, for instance : 

* a 1D signal with 500 points
* a single change point in the mean signal (around position 250)
* the same variance on both segments of the signal (the noise standard deviation added to the constant signal has a value of 1

We explore three scenarios : 

1. No change point
Now, if you consider no change point, then the cost function is computed on the whole signal and since you have a change in mean at around position 250, the computed variance on the complete signal is quite high. 
![no_change_point](https://user-images.githubusercontent.com/12909374/108983891-a379c700-768f-11eb-9f7c-a12e5ac92f45.png)

2. One change point, at an non-optimal position
Let say you choose that there is a change point at position 100. Then, the cost function on the first segment will be relatively low because the variance will be low on this segment. The cost function on the second segment will be on the other hand relatively high since the **real** change point in the mean at around position 250 introduces some additional variance.  
![non_optimal_change_point](https://user-images.githubusercontent.com/12909374/108984071-d02dde80-768f-11eb-8294-f85fba9d251f.png)

3. One change point, at optimal position 
Let say you choose that the change point is at its **real** location. Then you are fine. Indeed, on both the first and second segments, the computed variances on each segment do **not** include the **real** change point at around position 250. Therefore, on both segments the computed variance (hence the cost) will be relatively low. The sum of costs on your segments is then at the minimum value possible. 
![optimal_change_point](https://user-images.githubusercontent.com/12909374/108984241-fd7a8c80-768f-11eb-9722-36f1ac380577.png)

Please, find below a code sample you can run to illustrate this 🔽 

```python
import numpy as np
import matplotlib.pylab as plt
import ruptures as rpt

n, dim = 500, 1  # number of samples, dimension
n_bkps, sigma = 1, 1  # number of change points, noise standart deviation
signal, bkps = rpt.pw_constant(n, dim, n_bkps, noise_std=sigma, delta=(4, 5), seed=1234)

# Create cost object
c = rpt.costs.CostNormal().fit(signal)

# No change point
print("No change points:")
print(f"\tTotal cost:\t{c.error(0, 500):.2f}")
figure, axs = rpt.display(signal, bkps)
figure.show()

# 1 change point, at an non-optimal position
print("One non-optimal change points:")
print(f"\tCost on segment 1:\t{c.error(0, 100):.2f}")
print(f"\tCost on segment 2:\t{c.error(100, 500):.2f}")
print(f"\tTotal cost:\t{c.sum_of_costs([100, 500]):.2f}")
figure, axs = rpt.display(signal, bkps, [100, 500])
figure.show()

# 1 change point, at the optimal position
print("One optimal change points:")
print(f"\tCost on segment 1:\t{c.error(0, bkps[0]):.2f}")
print(f"\tCost on segment 2:\t{c.error(bkps[0], 500):.2f}")
print(f"\tTotal cost:\t{c.sum_of_costs(bkps):.2f}")
figure, axs = rpt.display(signal, bkps, bkps)
figure.show()

```

Hope I am clear enough and that you get the intuition behind it ? The logic is exactly the same for multidimensional signals and for several change points. 

Best,

Olivier
